// Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
//
// KVBlockChain commit service:
// Provides a service to commit KV transactions to the blockchain.
//

syntax = "proto3";
package com.digitalasset.kvbc;
option java_package = "com.digitalasset.kvbc";

message CommitRequest {
  bytes submission = 1;
  reserved 2; // was input_log_entries
  reserved 3; // was input_state
  string participant_id = 4;
  string correlation_id = 5;
  bytes span_context = 6;
  uint32 flags = 7;
  //TODO: Add signature over the submission
}

message CommitResponse {
  CommitStatus status = 1;

  // The block into which the transaction was committed.
  // Unset if commit failed.
  int64 block_id = 2;

  enum CommitStatus {
    OK = 0;
    ERROR = 1;
  }
}

message KVKey {
  bytes key = 1;
}


// FIXME(JM): Move to its own proto file
// The KVBC command
message Command {
  oneof cmd {
    ReadCommand read = 1;
    CommitRequest commit = 2;
  }
}

message ReadCommand {
  repeated KVKey reads = 1;
  string participant_id = 2;
  //TODO: Add signature over the reads
}

message CommandReply {
  oneof result {
    CommitResponse commit = 1;
    string read = 2; // TODO
    string error = 3;
  }
}

